exec/java-exec/src/test/java/org/apache/drill/test/TestGracefulShutdown.java

```diff
@@ -0,0 +1,248 @@
+          throw new RuntimeException("Failed : HTTP error code : "
+                  + conn.getResponseCode());
+        }
+        port++;
+      }
+      Thread.sleep(grace_period);
+      Collection<DrillbitEndpoint> drillbitEndpoints = cluster.drillbit().getContext()
+              .getClusterCoordinator()
+              .getOnlineEndPoints();
+      while(!listener.isDone()) {
+        Thread.sleep(10);
+      }
+      Assert.assertTrue(listener.isDone());
+      Assert.assertEquals(1,drillbitEndpoints.size());
+    }
+  }
+
+  /*
+   Test default shutdown through RestApi
+   */
+  @Test
+  public void testRestApiShutdown() throws Exception {
+
+    String[] drillbits = {"db1" ,"db2", "db3", "db4", "db5", "db6"};
+    ClusterFixtureBuilder builder = ClusterFixture.bareBuilder().withBits(drillbits).withLocalZk();
+    builder = enableWebServer(builder);
+    QueryBuilder.QuerySummaryFuture listener;
+    final String sql = "SELECT * FROM dfs."+testDirPath +" ORDER BY employee_id`";
+    try ( ClusterFixture cluster = builder.build();
+          final ClientFixture client = cluster.clientFixture()) {
+      Drillbit drillbit = cluster.drillbit("db1");
+      int port = drillbit.getContext().getConfig().getInt("drill.exec.http.port");
+      int grace_period = drillbit.getContext().getConfig().getInt("drill.exec.grace_period");
+      listener =  client.queryBuilder().sql(sql).futureSummary();
+      Thread.sleep(50000);
+      while( port < 8048) {
+        URL url = new URL("http://localhost:"+port+"/shutdown");
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+        conn.setRequestMethod("POST");
+        if (conn.getResponseCode() != 200) {
+          throw new RuntimeException("Failed : HTTP error code : "
+                  + conn.getResponseCode());
+        }
+        port++;
+      }
+      Thread.sleep(grace_period);
+      Thread.sleep(5000);
+      Collection<DrillbitEndpoint> drillbitEndpoints = cluster.drillbit().getContext()
+              .getClusterCoordinator().getAvailableEndpoints();
+      while(!listener.isDone()) {
+        Thread.sleep(10);
+      }
+      Assert.assertTrue(listener.isDone());
+      Assert.assertEquals(5,drillbitEndpoints.size());
+    }
+  }
+
+  public static File getTempDir(final String dirName) {
+    final File dir = Files.createTempDir();
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        FileUtils.deleteQuietly(dir);
+      }
+    });
+    File tempDir = new File(dir, dirName);
+    tempDir.mkdirs();
+    return tempDir;
+  }
+
+  private static File getTempFile(int file_num, File testDir) throws Exception {
+    return File.createTempFile("employee"+file_num, ".json", testDir);
+  }
+
+  private static void setupFile(File testDir, int file_num) throws Exception {
+    File destFile = getTempFile(file_num, testDir);
+    try (PrintWriter out = new PrintWriter(new FileWriter(destFile))) {
```

Comment:
Can this test use one of our many existing data files? Maybe the `customer` table on the class path?